### PR TITLE
Jsrt: External data support for DynamicObject [experiment PR]

### PR DESCRIPTION
**Pull request by obastemur**
Adds API below to ChakraCore Jsrt;

 -  JsObjectHasExternalData
 -  JsObjectGetExternalData
 -  JsObjectSetExternalData
 -  JsObjectSetExternalDataWithCallback

Removes API from ChakraCore common Jsrt;

- JsHasExternalData
- JsGetExternalData
- JsSetExternalData

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -63,6 +63,7 @@ test/benchmarks/*.dpl
 test/benchmarks/*.txt
 testout*
 packages.config
+lib/wabt/built/config.h
```
Comment (MSLaguana) [inline .gitignore]: Why should this header be excluded?

File: lib/Jsrt/Jsrt.cpp [hunk 2]
```diff
@@ -4764,7 +4714,7 @@ CHAKRA_API JsGetAndClearExceptionWithMetadata(_Out_ JsValueRef *metadata)
 
 CHAKRA_API JsHasOwnProperty(_In_ JsValueRef object, _In_ JsPropertyIdRef propertyId, _Out_ bool *hasOwnProperty)
 {
-    return ContextAPIWrapper<true>([&] (Js::ScriptContext *scriptContext, TTDRecorder& _actionEntryPopper) -> JsErrorCode {
+    return ContextAPIWrapper<JSRT_MAYBE_TRUE>([&] (Js::ScriptContext *scriptContext, TTDRecorder& _actionEntryPopper) -> JsErrorCode {
```
Comment (MSLaguana) [inline lib/Jsrt/Jsrt.cpp]: Looks like this is from one of your other changes; but I don't think it's directly relevant to this one.

Comment (obastemur) [reply lib/Jsrt/Jsrt.cpp]: @MSLaguana I saw this one was skipped and updated. Not sure I understood the problem?

**Comment by fatcerberus** [comment]
Does this change allowing setting a finalizer too?

**Comment by obastemur** [comment]
> Does this change allowing setting a finalizer too?

@fatcerberus Yes, `JsObjectSetExternalDataWithCallback`

**Comment by liminzhu** [comment]
Looks solid 👍 . A few feedback/questions,

1. Why put the old APIs in ChakraCommonWindows.h? Either leaving them as-is or deprecating them in ChakraCore (a.k.a. they show up in ChakraFull not here) makes more sense. I tend to leave them as-is, but that's ofc up for debate.

2. Can I put ext data on any JsValueRef, or just objects?

3. I assume every time `JsObjectSetExternalDataWithCallback` is called, it will replace any previous finalizer set on the object, including the one from JsCreateExternalObejct. A following `JsObjectSetExternalData` call should set the finalizer to nothing. Am I correct?

**Comment by obastemur** [comment]
@liminzhu I will put an external post to explain what is what in detail.

About JSRT; IMHO old `JsHasExternalData..` confusing and not needed.

**Comment by obastemur** [comment]
Definitely strange errors from CI

```
/Users/dotnet-bot/j/workspace/Microsoft_ChakraCore/release_1.7/static_osx_osx_debug_prtest/out/Debug/../../lib/Jsrt/ChakraCore.h:300:5: note: candidate function not viable: requires 4 arguments, but 5 were provided
```

Well... It's defined under ChakraCore.h:309.. not 300!..

**Comment by obastemur** [comment]
Looks like the CI issue is related to branch / base difference among the headers... Updated to latest again.

File: test/native-tests/test-char/sample.cpp [hunk 3]
```diff
@@ -3,23 +3,7 @@
 // Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
 //-------------------------------------------------------------------------------------------------------
 
-#include "ChakraCore.h"
```
Comment (obastemur) [inline test/native-tests/test-char/sample.cpp]: this is the one causing the CI fails.. (no nullptr is defined.) will update after review.

**Comment by obastemur** [comment]
Reminder; I'm just experimenting here. No promises. If your application depends on external data usage and current ChakraCore Jsrt is not helping, please let us know from https://github.com/Microsoft/ChakraCore/issues/3592 

Current version;

- External data support for non primitive JS types.
- If, needed in the future, same interface can host additional properties on a DynamicObject without increasing its initial size.
- No runtime / memory usage etc. change on anything that doesn't use external data.
- Perf cost is limited to couple of if/else call - bool function (vtable)
- Brings additional TypeHandler types. i.e. We had PathTypeHandler.. now we also have PathTypeHandlerWithExternal. The new interface is all about `HasExternalDataSupport()` virtual function is returning true. No additional logic update to original type handler.
- Centralize the usage of auxSlots and internally manages under DynamicObject
- Supports type handler sharing among the new object instances with external data.
- Added native tests to make sure things work -> https://github.com/Microsoft/ChakraCore/pull/3639/files#diff-f2ba368ddfd5ff52839d16738442a504

What else needs to be done;

- More testing!
- Reduce the amount of ConvertToExternalData calls. (latest update to PR cleared the hot paths already)

**Comment by obastemur** [comment]
@dotnet-bot test Copyright Check please

**Comment by fatcerberus** [comment]
Without this change, is there a way right now to have a native constructor create an object with external data?  If I return a new object from a constructor directly, then subclassing it doesn't work :(

**Comment by obastemur** [comment]
Was a good experiment and learned a lot. Thanks for the `thumbs up`s!

#3717 may help to implement such feature without changing the type system significantly

**Pull request closed without merging**